src/EventStore.Client.Streams/EventStoreClient.Read.cs

```diff
@@ -65,23 +71,115 @@ public partial class EventStoreClient {
 				cancellationToken);
 		}
 
-		private ReadStreamResult ReadStreamAsync(Direction direction,
-			string streamName, StreamPosition revision, long maxCount,
-			EventStoreClientOperationOptions operationOptions, bool resolveLinkTos = false,
-			UserCredentials? userCredentials = null, CancellationToken cancellationToken = default) {
-			return new ReadStreamResult(SelectCallInvoker, new ReadReq {
-					Options = new ReadReq.Types.Options {
-						ReadDirection = direction switch {
-							Direction.Backwards => ReadReq.Types.Options.Types.ReadDirection.Backwards,
-							Direction.Forwards => ReadReq.Types.Options.Types.ReadDirection.Forwards,
-							_ => throw InvalidOption(direction)
-						},
-						ResolveLinks = resolveLinkTos,
-						Stream = ReadReq.Types.Options.Types.StreamOptions.FromStreamNameAndRevision(streamName,
-							revision),
-						Count = (ulong)maxCount
+		/// <summary>
+		/// A class that represents the result of a read operation on a stream.
+		/// </summary>
+		public class ReadAllStreamResult : IAsyncEnumerable<ResolvedEvent> {
+			private readonly Channel<StreamMessage> _channel;
+
+			/// <summary>
+			/// 
+			/// </summary>
+			public Position? LastPosition { get; private set; }
+
+			/// <summary>
+			/// An <see cref="IAsyncEnumerable{StreamMessage}"/>.
+			/// </summary>
+			public IAsyncEnumerable<StreamMessage> Messages {
+				get {
+					return GetMessages();
+
+					async IAsyncEnumerable<StreamMessage> GetMessages() {
+						await foreach (var message in _channel.Reader.ReadAllAsync().ConfigureAwait(false)) {
```

Comment:
some (small?) risk here that the user might enumerate `Messages`
- twice in series expecting to be able to read them twice (but second time would be empty)
- twice concurrently expecting to be able to distribute the messages to multiple handlers (but would violate the singlereader assumption on the chanel)

not sure if this would just need documenting, or more protection than that

Reply:
Guard it with a boolean and throw an exception if the user tries to call this twice. Also, update the documentation to call this out